src/Groups/directproducts.jl

```diff
@@ -0,0 +1,486 @@
+################################################################################
+#
+#  Managing direct products
+#
+################################################################################
+
+export
+    as_matrix_group,
+    as_perm_group,
+    as_polycyclic_group,
+    cartesian_power,
+    direct_product,
+    embedding,
+    factorofdirectproduct,
+    homomorphism_of_semidirect_product,
+    isfull_direct_product,
+    isfull_semidirect_product,
+    isfull_wreath_product,
+    number_of_factors,
+    projection,
+    semidirect_product,
+    sub,
+    wreath_product,
+    write_as_full
+
+
+"""
+    direct_product(L::AbstractVector{<:GAPGroup})
+    direct_product(L::GAPGroup...)
+Return the direct product of the groups in the vector `L`.
+"""
+function direct_product(L::AbstractVector{<:GAPGroup})
+   X = GAP.Globals.DirectProduct(GAP.julia_to_gap([G.X for G in L]))
+   n = length(L)
+   return DirectProductOfGroups(X,L,n,X,true)
+end
+
+function direct_product(L::GAPGroup...)
+   return direct_product([x for x in L])
+end
+
+"""
+    number_of_factors(G::DirectProductOfGroups)
+Return the number of factors of `G`.
+"""
+number_of_factors(G::DirectProductOfGroups) = G.n
+
+"""
+    cartesian_power(G::T, n::Int)
+Return the direct product of `n` copies of `G`, of type `T`.
+"""
+function cartesian_power(G::T, n::Base.Integer) where T <: GAPGroup
+   L = [G for i in 1:n]
+   return direct_product(L)
+end
+
+"""
+    factorofdirectproduct(G::DirectProductOfGroups, j::Int)
+Return the `j`-th factor of `G`.
+"""
+function factorofdirectproduct(G::DirectProductOfGroups, j::Base.Integer)
+   if j in 1:G.n return G.L[j]
+   else throw(ArgumentError("index not valid"))
+   end
+end
+
+"""
+    as_perm_group(G::DirectProductOfGroups)
+If `G` is direct product of permutations groups, return `G` as permutation group.
+"""
+function as_perm_group(G::DirectProductOfGroups)
```

Comment:
I don't think this is so useful, is it? At least I can't think of too many applications. Instead, I would suggest two things (which are not mutually exclusive, i.e. we may want to do both):

1. We could have a function `inner_direct_product` or so which takes as arguments groups that are all of the same type, and then returns a direct product *of the same type*, together with lists of embeddings and projections (which can be turned on/off via optional named arguments; I think @fieker mentioned that this kind of stuff is already done elsewhere, perhaps in Hecker)

2. We add functions `isomorphism_perm_group`, `isomorphism_fp_group` etc. which essentially wrap `GAP.Globals.IsomorphismPermGroup`. Variant: `isomorphic_perm_group` which returns a perm group and an isomorphism. Variant of the variant: call it `as_perm_group` (as you did here).
```julia
function isomorphic_perm_group(G::GAPGroup)
  iso = GAP.Globals.IsomorphismPermGroup(G.X) # TODO: catch errors?
  iso == GAP.Globals.fail && error("could not convert group into a permutation group")
  H = Image(iso)
  return TODO_as_GAP_group_object(H), TODO_as_hom_obj(iso)
end
```

Reply:
Yes, I wrote this function because you (or Claus, I don't remember) suggested that direct products of groups of the same type `T` should have also type `T` (where `T` = `PermGroup`, `PcGroup` or `MatrixGroup`).
The two suggestions you made face the issue more exhaustively, I can do them easily. In such a case `as_perm_group(G)` would be just equivalent to `isomorphic_perm_group(G)[1]`, so it can be deleted.

Reply:
>     1. We could have a function `inner_direct_product` or so which takes as arguments groups that are all of the same type, and then returns a direct product _of the same type_, together with lists of embeddings and projections (which can be turned on/off via optional named arguments; I think @fieker mentioned that this kind of stuff is already done elsewhere, perhaps in Hecker)

About this last function: if the direct product `G` is created as a group of type `T` (where `T` = `PermGroup`, `PcGroup`, etc.), on this group are no longer defined peculiar functions for direct products, such as `embedding`, `projection`, or the instruction that allows to create the element `(x1,x2,x3)` of the direct product simply typing `G(x1,x2,x3)`. Do we make these functions available again?